### DISCUSSION TITLE
Blocking characteristics

### DISCUSSION THREAD

**Discussion by agersant** [vincent-herlemont/native_db#240] [Q&A] [answered]
Hello! Would it be possible to get some general idea of the expected performance characteristics of transaction operations?

More precisely, I would like to know:
- In what circumstances do transaction operations (get, insert, remove, commit, etc.) end up performing disk I/O on the thread they were called?
- Is the above answer any different if the amount of managed data is below the [cache_size](https://docs.rs/native_db/latest/src/native_db/database_builder.rs.html#74-77)?
- How frequently is data persisted to disk, and what guarantees are there in case of unexpected interruptions (eg. power loss)?
- I see there is some `tokio` support in `Cargo.toml` but it's not clear to me what enabling it would do.

**Comment by vincent-herlemont** [comment, marked as answer]
Native DB entirely delegates the management of storage to the [`redb`](https://github.com/cberner/redb) backend (database) and it's necessary to refer to this documentation for details [redb/design](https://github.com/cberner/redb/blob/master/docs/design.md). Native DB can be seen as an overlay to `redb`. However, here are some brief responses:

> In what circumstances do transaction operations (get, insert, remove, commit, etc.) end up performing disk I/O on the thread they were called?

Transaction operations perform disk I/O when accessing data not in cache or when writing changes to disk during a commit.

> Is the above answer any different if the amount of managed data is below the [cache_size](https://docs.rs/native_db/latest/src/native_db/database_builder.rs.html#74-77)?

Yes, if all data fits within the cache, read operations may avoid disk I/O, but writes still require disk I/O during commits to persist changes.

> How frequently is data persisted to disk, and what guarantees are there in case of unexpected interruptions (eg. power loss)?

Data is persisted to disk during commit operations, and `redb` ensures data integrity during unexpected interruptions by using checksums, atomic writes, and recovery mechanisms.

---

> I see there is some `tokio` support in Cargo.toml but it's not clear to me what enabling it would do.

For the moment, `tokio` for Native DB has nothing to do with storage but is simply the support for channels used to listen to database events useful for real-time applications (see [`watch`](https://docs.rs/native_db/latest/native_db/struct.Database.html#method.watch)).

**Reply by agersant** [reply]
Thank you very much!
